aiohttp/cookiejar.py

```diff
@@ -175,7 +175,9 @@ def update_cookies(self,
                 if expires:
                     expire_time = self._parse_date(expires)
                     if expire_time:
-                        self._expire_cookie(expire_time.timestamp(),
+                        diff = (expire_time.timestamp() -
```

Comment:
From my understanding the first parameter of `_expire_cookie` is called `when` and it is an absolute time (seconds since UXIN epoch), not a delta with the current time.
I like to keep the parameter meaning.

Reply:
I don't believe that's a true assumption for `_expire_cookie`. Look at line 168 where it's used for max-age expiration. It's a time relative to the loop_time().  

The first parameter is the sum of the `self._loop.time()` + seconds left from max-age. It's a relative time and not tied to UNIX epoch because `loop.time()` is [isn't tied to the UNIX epoch](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.time).  Because `loop.time()` isn't tied to the UNIX epoch, the correct answer is to pass some value <= 0 which will cause an immediate expiration.  You could include some additional math functions to force it to be 0 if you don't want it to be negative, but that isn't necessary.

Reply:
If it's supposed to be `time.time()`, then I can easily change my edit to `time.time()+diff` for consistency.  Will that resolve your concern?  However, line 86 in `do_expiration()` will break because you're comparing `loop.time` which will always be much smaller than `time.time()`.  To resolve that I could just do a `loop.time() + diff`, which will also typically be a very negative number (again since unix epochs are so much larger).

For the test suite, I still need an answer on whether CookieJar should allow storage of expired cookies as that is a core assumption in the test suite.

Reply:
I mean using time.time() consistently in `CookieJar` for representing absolute time looks like a proper fix. It doesn't need `loop.time()` at all, at least seems so.

Reply:
Yes, I think that looks right. Do you want me to make that switch throughout?

Reply:
It can be done with datetime unless you want to import time too.

Reply:
Please do. `datetime` is even better, we never mix absolute and relative time again.

Reply:
Ok, what about the test suite assumption?

Reply:
Do you mean tests that become fail after the code change?
Tests should be updated.

Reply:
There are indeed more issues here. If `datetime` is desired then I'd rework the time components in the whole class and use `datetime.datetime` instances with UTC time throughout.

* `MAX_TIME` should be `datetime.datetime.max.replace(tzinfo=datetime.timezone.utc)`
* `self._next_expiration = ceil(self._loop.time())` in the `CookieJar` `__init__` `_clear()`  and  should be set to `datetime.datetime.now(datetime.timezone.utc).replace(microsecond=0) + datetime.timedelta(seconds=0)` (so microseconds cleared and a second added to align with the next whole second).
* The `_do_expirations()` method should switch to use `now = datetime.datetime.now(datetime.timezone.utc)` and `self._next_expiration = next_expiration.replace(microsecond=0) + datetime.timedelta(seconds=1)`
* alter `_expire_cookie` to make `when` a `datetime` instance:

    ```python
    def _expire_cookie(self, when: datetime.datetime, domain: str, name: str) -> None:
        self._next_expiration = min(self._next_expiration, when)
        self._expirations[(domain, name)] = when
    ```
* Change the `max-age` handling to pass in `datetime.datetime.now(timezone.utc) + delta_seconds)`.

Reply:
Ok I've implemented this; time to work on the tests and make sure I haven't broken anything.